### PR TITLE
Remove the ability to update datafeed's job_id.

### LINKED ISSUE #44616: [ML] Remove the ability to move datafeeds between jobs
After #44615 has been implemented the ability to move datafeeds between jobs should be removed from the master branch.  As anomaly detection jobs and datafeeds have become more closely coupled this is undesirable and adds complexity.

The update datafeed REST endpoint should be changed to report an error if it receives a request containing `job_id`.  If an 8.x node receives a request over wire transport that contains `job_id` then it should refuse to process the request and return an error response.

Removal of the ability to move a datafeed to a different job should then open up the possibility to simplify other code.  For example there will be no need for 8.x to consider deleting datafeed timing stats when moving a datafeed to a different job.  There may be other places where the code can be simplified too.

### PR DESCRIPTION
**Pull request by przemekwitek**
This PR is a followup to https://github.com/elastic/elasticsearch/pull/44691. It disables the possibility to update job_id on a datafeed.
DatafeedUpdate.jobId field can be set in the _update request as long as it is the same as the current value of DatafeedConfig.jobId.

Closes https://github.com/elastic/elasticsearch/issues/44616

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java [hunk 1]
```diff
@@ -298,6 +293,9 @@ public DatafeedConfig apply(DatafeedConfig datafeedConfig, Map<String, String> h
 
         DatafeedConfig.Builder builder = new DatafeedConfig.Builder(datafeedConfig);
         if (jobId != null) {
+            if (datafeedConfig.getJobId() != null && datafeedConfig.getJobId().equals(jobId) == false) {
+                throw ValidationException.withError(ERROR_MESSAGE_ON_JOB_ID_UPDATE);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java]: A possible problem with using `ValidationException` here is that this method is being called on the master node but the original request is likely to have been sent to some other node.  This exception will get wrapped in a `RemoteTransportException` when it is sent back to the coordinating node.  There is logic to determine the "root cause" of an exception, and the rule is that it's the most deeply nested exception that inherits from `ElasticsearchException`.  Since `RemoteTransportException` inherits from `ElasticsearchException` and `ValidationException` doesn't, this means the `RemoteTransportException` will be reported as the root cause.

Exceptions that are root causes but do not inherit from `ElasticsearchException` (such as `IllegalArgumentException` or `ValidationException`) should only be used in places where they'll be thrown on the coordinating node.  Any exceptions that are root causes and will be thrown on a different node must extend `ElasticsearchException`.

TL;DR it's probably better to use a `ElasticsearchStatusException` here, which you can do by calling `org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper.badRequestException`.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/datafeed/DatafeedUpdate.java [hunk 2]
```diff
@@ -287,6 +287,7 @@ public Builder(DatafeedUpdate config) {
             this.delayedDataCheckConfig = config.delayedDataCheckConfig;
         }
 
+        @Deprecated
```
Comment (droberts195) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/datafeed/DatafeedUpdate.java]: I think this annotation should be in 7.4 and the method removed entirely in 8.x.

When we decided to allow `job_id` to still be specified as long as it matched the current `job_id` this was for the get-modify locally-update workflow.  But that only makes sense if some code is taking a whole datafeed config from the get API and sending it back with modifications to the update API.  This method is setting the `job_id` individually, and if somebody is doing that they might as well delete the call from their code when upgrading to 8.x.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java (hunk 1 above)
Comment (przemekwitek) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java]: Done.

Thanks for detailed explanation of exception handling!

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by przemekwitek** [comment]
run elasticsearch-ci/1
